### ISSUE TITLE
Support mirror/recursive COPY

### ISSUE THREAD

**Issue by Nirmal4G** [moby/moby#39530]
**Description**

The _**COPY**_ command, presently cannot copy/mirror the source directory structure (or pathspec) in the target image.

**Steps to reproduce the issue:**
- Use `COPY Source/**/*.ext Target/` in your Dockerfile

**Describe the results you received:**
Every file matching the path spec is copied into the `Target` folder without the files' original directory structure/layout.

**Describe the results you expected:**
Files should be copied along with the directory structure/layout preserved.

**Solution to the issue:**
An option should be presented as a secondary switch to `COPY` command to enable recursive/mirror copying. I recommend to have

```Dockerfile
COPY [-r|--recursive] <source-path-spec> <target-path>
COPY [-m|--mirror] <source-path-spec> <target-path>
```

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        6247962
 Built:             Sun Feb 10 04:12:31 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.24)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 04:28:48 2019
  OS/Arch:          windows/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.09.2
Storage Driver: windowsfilter
 Windows: 
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd gelf json-file local logentries splunk syslog
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 18362 (18362.1.amd64fre.19h1_release.190318-1202)
Operating System: Windows 10 Pro Version 1903 (OS Build 18362.239)
OSType: windows
Architecture: x86_64
CPUs: 8
Total Memory: 31.97GiB
Name: WinX-AIO
ID: NTRV:UDM2:QK6B:BLI5:RM7C:7US5:SZ2V:LW2Q:HBPN:DE4A:I566:CZ6N
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 27
 System Time: 2019-07-16T01:28:15.2388227+05:30
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

**Comment by justincormack** [comment]
possibly `COPY --cp` to indicate it is like the cp command?

**Comment by thaJeztah** [comment]
Also see https://github.com/moby/moby/issues/38710#issuecomment-462734118, https://github.com/moby/moby/issues/35639, and (mentioning that) https://github.com/moby/buildkit/issues/396

There's some other issues with `COPY` though (also see https://github.com/moby/moby/issues/15858), and I'm not sure we'd be able to address all of them (and if we can implement it to be acting exactly the same as `cp` (which also seems to have different implementations).

With buildkit, it's possible to use `RUN --mount`, which would allow you to do (see the [experimental syntax](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md#experimental-syntaxes));

```
RUN --mount=target=Source cp Source/**/*.ext Target/
```

([among other things](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md#experimental-syntaxes)

Unfortunately, BuildKit is not yet available for native Windows containers, so not an option here.

**Comment by Nirmal4G** [comment]
Still as you know, these behaviors has been asked a lot from the community. It's time that we address these issues.

Do you guys have a design repo for these projects?
If so, please point me there, I can create a high level proposals for some of the issues I've faced over time when using docker!

**Comment by dustinlacewell** [comment]
https://github.com/moby/moby/issues/15858 is closed but this is open, yet all the discussion is on the other. There needs to be a single issue, with the resolve to keep it open, to track how COPY's useless globbing mechanics need to be desperately fixed after _**all these years**_.

**Comment by ruffsl** [comment]
Should there be a `--parents` flag, similar to cp to specify preserving the directory structure?

> --parents
> use full source file name under DIRECTORY
https://linux.die.net/man/1/cp

**Comment by ruffsl** [comment]
> With buildkit, it's possible to use `RUN --mount`, which would allow you to do (see the [experimental syntax](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md#experimental-syntaxes));
> 
> ```
> RUN --mount=target=Source cp Source/**/*.ext Target/
> ```

@thaJeztah : Unforchunetly, even when using build kit, this doesn't change the fact that the caching awareness for the RUN directive is not as precise as that of COPY. Thus while the experimental example demonstrates the using glob style copying to include an common pattern of files, it can't be used to help preserve the build cache as selectively.

I.e. currently the directive you've written leaves the `from` options unset, e.g. the build stage or image name for the root of the source, which defaults to the build context. Thus if any file changes in the build context, that instance of the RUN directive will break from its cache, regardless if the globed files it copies from the build context have changed themselves or not.

Additionally, given the default shell is `/bin/sh`, the `**` syntax from globstar would be unavailable.

In summary, the buildkit example Dockerfile below would not have the equivalent cache robust persistence nor sensitivity as that which was proposed in my comment here: https://github.com/moby/moby/issues/15858#issuecomment-462017830

``` Dockerfile
# syntax = docker/dockerfile:1.0-experimental

# copy project dependency metadata
RUN --mount=target=/tmp/src \
    cd /tmp/src && \
    find ./ -name "package.xml" | \
    xargs cp --parents -t /opt/ws/

# install step that fetches unsatisfied dependency
RUN dependency_manager install --workspace /opt/ws/

# copy the rest of the project's code
COPY ./ /opt/ws/

# compile code with cached dependencies
RUN build_tool build --workspace /opt/ws/
```

**Comment by entrptaher** [comment]
This is really painful to see and update.

```
COPY  ./packages/shared-backend/package.json ./packages/shared-backend/package.json
COPY  ./packages/shared-dashboard/package.json ./packages/shared-dashboard/package.json
COPY  ./packages/shared-extractor/package.json ./packages/shared-extractor/package.json
COPY  ./packages/shared-queue-worker/package.json ./packages/shared-queue-worker/package.json
COPY  ./packages/shared-shared/package.json ./packages/shared-shared/package.json
COPY  ./packages/shared-chain-of-command/package.json ./packages/shared-chain-of-command/package.json
COPY  ./packages/shared-driver/package.json ./packages/shared-driver/package.json
COPY  ./packages/shared-queue-publisher/package.json ./packages/shared-queue-publisher/package.json
COPY  ./packages/shared-runner/package.json ./packages/shared-runner/package.json
```

I have 100+ packages on a repo, and every little takes forever no matter how much I optimize it.

The following only works inside CMD, not on the build process, and worse, the container is removed immediately.
```
CMD cp --parents /packages/*/package.json /app && ls -la /app/packages 
```

This feature looks like it will help us save tons of time on the build process.

**Comment by AkihiroSuda** [comment]
SGTM
cc @tonistiigi @tiborvass

**Comment by feverpitch-tom** [comment]
+1 for this functionality being desirable

**Comment by Nirmal4G** [comment]
Any status on this?

**Comment by mackeyja92** [comment]
Any update here?

**Comment by tonistiigi** [comment]
@ruffsi You just need to do a `COPY` of the files with multi-stage after you have filtered out the files you need with `cp --parents`. Then if files change the filter step will run again but if none of the files that actually match the filter have changed the cache picks up again.

**Comment by ruffsl** [comment]
@tonistiigi , yep. Since my original comments here I have gone onto use this multi-stage copy caching approach to arbitrarily filter what files are selectively used in busting consecutive layers in following stages. Some examples for other readers:

https://github.com/rust-lang/cargo/issues/2644#issuecomment-614257807

It's a bit more verbose as build pattern, but it's also lot more flexible given you can use Turing complete scripts to filter or pre-process files from the build context. Although in most cases you may be just copying the entire build context, it'll frequently bust the cacher stage, needing the stage to be rebuilt, regardless if that cacher stage does or does not bust image layers in following stages.

https://github.com/docker-library/docs/pull/1381/files#diff-9bc9da827253d9e984e1d95aeb068ae1R68

If you have any suggestions to improve the pattern, I'd be happy to incorporate it into some docs.

**Comment by yehonatanz** [comment]
Are there plans to address this issue in the foreseeable future?
This will greatly benefit cache utilization in multi-package repos, which are a common occurrence these days.

Right now this is a real problem for us and we resort to auto-generation of the `Dockerfile`, which is far from optimal.
